### PR TITLE
Handle empty offsets buffer (#1824)

### LINKED ISSUE #1824: Invalid Value Offsets Slice For Empty Variable Length Arrays
**Describe the bug**

I don't know if this was a change or it has always been this way, but we permit zero-length ListArray to have no offsets. In #1620 we explicitly decided against enforcing a single value.

Unfortunately the logic for value_offsets, duplicated in ListArray, StringArray, BinaryArray, etc...

```
unsafe {
    std::slice::from_raw_parts(
        self.value_offsets.as_ptr().add(self.data.offset()),
        self.len() + 1,
    )
}
```

Does not take this into account

**To Reproduce**

```
#[test]
fn test_string_array_empty_offsets() {
    let data = ArrayDataBuilder::new(DataType::Utf8)
        .len(0)
        .add_buffer(Buffer::from([]))
        .add_buffer(Buffer::from([]))
        .build()
        .unwrap();
    let array = GenericStringArray::<i32>::from(data);
    assert_eq!(array.value_offsets().len(), 0); // FAILS with 1 != 0
}
```

**Expected behavior**

Given the decision to support empty offsets, I think we should always return an empty slice for zero length array types, so that the behaviour is consistent.

**Additional context**

Discovered whilst working on #1811, so already getting returns from that investment :grin:

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #1824

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

No

### CODE REVIEW

File: arrow-array/src/array/list_array.rs [hunk 1]
```diff
@@ -43,6 +43,17 @@ impl OffsetSizeTrait for i64 {
     const PREFIX: &'static str = "Large";
 }
 
+/// Returns a slice of `OffsetSize` consisting of a single zero value
+#[inline]
+pub(crate) fn empty_offsets<OffsetSize: OffsetSizeTrait>() -> &'static [OffsetSize] {
+    static OFFSET: &[i64] = &[0];
```
Comment (tustvold) [inline arrow-array/src/array/list_array.rs:49]: This slightly odd construction is necessary to workaround the fact Rust doesn't allow statics that depend on type parameters

**Comment by jhorstmann** [comment]
This looks like a nice trick :+1:

File: arrow-array/src/array/binary_array.rs [hunk 2]
```diff
@@ -845,4 +852,22 @@ mod tests {
...
+    #[test]
+    fn test_empty_offsets() {
+        let string = BinaryArray::from(
+            ArrayData::builder(DataType::Binary)
+                .buffers(vec![Buffer::from(&[]), Buffer::from(&[])])
+                .build()
+                .unwrap(),
+        );
+        assert_eq!(string.value_offsets(), &[0]);
+        let string = LargeBinaryArray::from(
```
Comment (alamb) [inline arrow-array/src/array/binary_array.rs:865]:
```suggestion
        let string = LargeBinaryArray::from(
        assert_eq!(string.len(), 0);
```

I suggest you add a test for `len()` as described in the original bug report.

This comment applies to all the tests added in this PR

**Review by alamb — APPROVED** [review]
Looks like a nice improvement to me -- thanks @tustvold 

I wondered if maybe `StructArray` needed similar treatment but it didn't look like it with my initial review

https://github.com/apache/arrow-rs/blob/master/arrow-array/src/array/struct_array.rs#L103-L111

**Pull request merged**